### PR TITLE
fixes permissions on work package context menu

### PR DESCRIPTION
**Pull request by ulferts**
Fixes: https://www.openproject.org/work_packages/8860

### CODE REVIEW

File: app/controllers/api/experimental/concerns/can.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+          @cache ||= Hash.new do |hash, work_package|
+            # copy checks for the move_work_packages permission. This makes
+            # sense only because the work_packages/moves controller handles
+            # copying multiple work packages.
+            hash[work_package] = {
+              :edit => edit_allowed?(work_package),
+              :log_time => log_time_allowed?(work_package),
+              :move => move_allowed?(work_package),
+              :copy => move_allowed?(work_package),
+              :duplicate => copy_allowed?(work_package), # duplicating is another form of copying
```
Comment (tilt) [inline app/controllers/api/experimental/concerns/can.rb:65]: looks a bit confusing to me here that the `duplicate` permission check is transitively forwarded to `move` via `copy`

File: app/controllers/api/experimental/concerns/can.rb [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            hash[project] = user.allowed_to?(:add_work_packages, project)
+          end
+
+          @add_cache[work_package.project]
+        end
+
+        def type_active_in_project?(work_package)
+          @type_active_cache ||= Hash.new do |hash, project|
+            hash[project] = project.types.pluck(:id)
+          end
```
Comment (tilt) [inline app/controllers/api/experimental/concerns/can.rb:118]: nice ruby tricks

**Comment by tilt** [comment]
@ulferts thanks for tidying up and the thoughtful design

File: app/controllers/api/experimental/concerns/can.rb (hunk 1 above)
Comment (ulferts) [inline app/controllers/api/experimental/concerns/can.rb:65]: I have to agree that this is confusing and it made me stumble. 
However this is the way it is for now until we clean up the routes. But there is no transitivity here from duplicate to move. Duplicate checks for the `add_work_packages` permission and whether the type of the work package copied from is active in the project of the copied work  package. Copy simply uses the `move_allowed?` check so it relies on the user to have the `move_work_packages` permission.

I am not saying that this makes sense but it mirrors the context menu as it used to be.

**Pull request merged**